chasm/lib/scheduler/invoker_tasks.go

```diff
@@ -0,0 +1,603 @@
+
+	// Because the catchup window doesn't apply to a manual start, pick a custom
+	// execution deadline before timing out a start.
+	manualStartExecutionDeadline = 1 * time.Hour
+
+	// Upper bound on how many times starting an individual buffered action should be retried.
+	InvokerMaxStartAttempts = 10 // TODO - dial this up/remove it
+)
+
+var (
+	errRetryLimitExceeded       = errors.New("retry limit exceeded")
+	_                     error = &rateLimitedError{}
+)
+
+func NewInvokerExecuteTaskExecutor(opts InvokerTaskExecutorOptions) *InvokerExecuteTaskExecutor {
+	return &InvokerExecuteTaskExecutor{
+		InvokerTaskExecutorOptions: opts,
+	}
+}
+
+func NewInvokerProcessBufferTaskExecutor(opts InvokerTaskExecutorOptions) *InvokerProcessBufferTaskExecutor {
+	return &InvokerProcessBufferTaskExecutor{
+		InvokerTaskExecutorOptions: opts,
+	}
+}
+
+func (e *InvokerExecuteTaskExecutor) Validate(
+	_ chasm.Context,
+	invoker *Invoker,
+	_ chasm.TaskAttributes,
+	_ *schedulespb.InvokerExecuteTask,
+) (bool, error) {
+	// If another execute task already happened to kick everything off, we don't need
+	// this one.
+	eligibleStarts := invoker.getEligibleBufferedStarts()
+	valid := len(invoker.GetTerminateWorkflows())+
+		len(invoker.GetCancelWorkflows())+
+		len(eligibleStarts) > 0
+	return valid, nil
+}
+
+func (e *InvokerExecuteTaskExecutor) Execute(
+	ctx context.Context,
+	invokerRef chasm.ComponentRef,
+	_ chasm.TaskAttributes,
+	_ *schedulespb.InvokerExecuteTask,
+) error {
+	var result executeResult
+
+	var invoker *Invoker
+	var scheduler *Scheduler
+
+	// Read and deep copy returned components, since we'll continue to access them
+	// outside of this function (outside of the MS lock).
+	_, err := chasm.ReadComponent(
+		ctx,
+		invokerRef,
+		func(i *Invoker, ctx chasm.Context, _ any) (struct{}, error) {
+			invoker = &Invoker{
+				InvokerInternal: common.CloneProto(i.InvokerInternal),
+			}
+
+			s, err := i.Scheduler.Get(ctx)
+			if err != nil {
+				return struct{}{}, err
+			}
+			scheduler = &Scheduler{
+				SchedulerInternal:  common.CloneProto(s.SchedulerInternal),
+				cacheConflictToken: s.cacheConflictToken,
+				compiledSpec:       s.compiledSpec,
+			}
+
+			return struct{}{}, nil
+		},
+		nil,
+	)
+	if err != nil {
+		return fmt.Errorf("%w: %w",
+			serviceerror.NewInternal("Failed to read component"),
+			err)
+	}
+
+	logger := newTaggedLogger(e.BaseLogger, scheduler)
+
+	// Terminate, cancel, and start workflows. The result struct contains the
+	// complete outcome of all requests executed in a single batch.
+	ictx := e.newInvokerTaskExecutorContext(ctx, scheduler)
+	result = result.Append(e.terminateWorkflows(ictx, logger, scheduler, invoker.GetTerminateWorkflows()))
+	result = result.Append(e.cancelWorkflows(ictx, logger, scheduler, invoker.GetCancelWorkflows()))
+	sres, startResults := e.startWorkflows(ictx, logger, scheduler, invoker.getEligibleBufferedStarts())
+	result = result.Append(sres)
```

Comment:
Could you please document some of what you said in this comment here?

> Can we run as many of these requests concurrently? Is there a reason to run sequentially?

The way this is structured doesn't suggest it, but ProcessBuffer won't ever actually have workflows to terminate, cancel, and start, simultaneously in the same run (only 1 of the 3). The way terminate/cancel work is that Execute makes the CancelWorkflow/TerminateWorkflow call, which triggers the workflow watcher callback mechanism, updating the Scheduler's RunningWorkflows list. ProcessBuffer looks at that list when determining whether to issue cancels, terminates, or starts, so once the callback finishes its transition, ProcessBuffer will begin to issue starts instead of cancels.